Memory leak/corruption with experimental ui_widget `observe()` function

## Bevy version and features

```
bevy = { version = "0.17.3", default-features = true, features = [
  "experimental_bevy_ui_widgets",
] }
```

## \[Optional\] Relevant system information

**Rust/Cargo version:**
rustc 1.93.0-nightly (1be6b13be 2025-11-26)
binary: rustc
commit-hash: 1be6b13be73dc12e98e51b403add4c41a0b77759
commit-date: 2025-11-26
host: x86_64-pc-windows-msvc
release: 1.93.0-nightly
LLVM version: 21.1.5



## What you did

Using the `observe()` function from the experimental ui_widget feature to spawn an observer with moved values.

Minimal reproduction example:

```rust
use std::time::Duration;

use bevy::{prelude::*, time::common_conditions::on_timer, ui_widgets::observe};

#[derive(Component)]
struct MyEntity;

#[derive(EntityEvent)]
struct MyEvent(Entity);

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, spawn_entity)
        .add_systems(
            Update,
            send_event_every_second.run_if(on_timer(Duration::from_secs(1))),
        )
        .run();
}

fn spawn_entity(mut commands: Commands) {
    let dynamic_string = String::from("Hello, Bevy!");
    info!("Dynamic string before closure: '{}'", dynamic_string);
    commands.spawn((
        MyEntity,
        Name::new("My Spawned Entity"),
        observe(move |_trigger: On<MyEvent>| {
            info!("Dynamic string inside closure: '{}'", dynamic_string);
        }),
    ));
}

fn send_event_every_second(query: Query<Entity, With<MyEntity>>, mut commands: Commands) {
    for entity in query.iter() {
        commands.trigger(MyEvent(entity));
        info!("Sent MyEvent to entity {:?}", entity);
    }
}
```

## What went wrong

I expected the string within the closure to be the same as outside, i.e. "Hello, Bevy!".
But the string inside the closure are always random bytes.
Sometimes part of the real string, sometimes random string, sometimes the program panics with "invalid value for char"

Output of some example runs of the reproduction code above:
```
2025-12-16T16:27:29.314699Z  INFO bevy_winit::system: Creating new window minimal_repro (0v0)
2025-12-16T16:27:29.386622Z  INFO minimal_repro: Dynamic string before closure: 'Hello, Bevy!'
2025-12-16T16:27:30.391179Z  INFO minimal_repro: Sent MyEvent to entity 12v0
2025-12-16T16:27:30.391355Z  INFO minimal_repro: Dynamic string inside closure: 'copy_bind_gr'
```

```
2025-12-16T16:26:27.367681Z  INFO bevy_winit::system: Creating new window minimal_repro (0v0)
2025-12-16T16:26:27.427773Z  INFO minimal_repro: Dynamic string before closure: 'Hello, Bevy!'
2025-12-16T16:26:28.431345Z  INFO minimal_repro: Sent MyEvent to entity 12v0
2025-12-16T16:26:28.431545Z  INFO minimal_repro: Dynamic string inside closure: '☻☺'
```

Also despawning this observer sometimes results in a heap corruption panic.

## Additional information

I guess there is a memory leak within the observer function in combination with a move closure.

Without the move keyword the program (correctly) won't compile, even if the value is cloned inside the closure:
```
error[E0373]: closure may outlive the current function, but it borrows `dynamic_string`, which is owned by the current function
  --> minimal_repro\src\main.rs:28:17
   |
28 |         observe(|_trigger: On<MyEvent>| {
   |                 ^^^^^^^^^^^^^^^^^^^^^^^ may outlive borrowed value `dynamic_string`
29 |             let inside_string = dynamic_string.clone();
   |                                 -------------- `dynamic_string` is borrowed here
   |
```

--- a/crates/bevy_ui_widgets/src/observe.rs
+++ b/crates/bevy_ui_widgets/src/observe.rs
@@ -2,7 +2,7 @@
 // It is certainly a useful thing to have.
 #![expect(unsafe_code, reason = "Unsafe code is used to improve performance.")]
 
-use core::marker::PhantomData;
+use core::{marker::PhantomData, mem};
 
 use bevy_ecs::{
     bundle::{Bundle, DynamicBundle},
@@ -48,18 +48,21 @@ impl<E: EntityEvent, B: Bundle, M, I: IntoObserverSystem<E, B, M>> DynamicBundle
 
     #[inline]
     unsafe fn get_components(
-        _ptr: bevy_ecs::ptr::MovingPtr<'_, Self>,
+        ptr: bevy_ecs::ptr::MovingPtr<'_, Self>,
         _func: &mut impl FnMut(bevy_ecs::component::StorageType, bevy_ecs::ptr::OwningPtr<'_>),
     ) {
-        // SAFETY: Empty function body
+        // SAFETY: We must not drop the pointer here, or it will be uninitialized in `apply_effect`
+        // below.
+        mem::forget(ptr);
     }
 
     #[inline]
     unsafe fn apply_effect(
-        ptr: bevy_ecs::ptr::MovingPtr<'_, core::mem::MaybeUninit<Self>>,
+        ptr: bevy_ecs::ptr::MovingPtr<'_, mem::MaybeUninit<Self>>,
         entity: &mut bevy_ecs::world::EntityWorldMut,
     ) {
-        // SAFETY: `get_components` does nothing, value was not moved.
+        // SAFETY: The pointer was not dropped in `get_components`, so the allocation is still
+        // initialized.
         let add_observer = unsafe { ptr.assume_init() };
         let add_observer = add_observer.read();
         entity.observe(add_observer.observer);

